cred/src/main/java/org/opencadc/cred/GetCertAction.java

```diff
@@ -1,96 +1,310 @@
+import java.security.AccessControlException;
+import java.security.KeyPair;
+import java.security.KeyPairGenerator;
+import java.security.NoSuchAlgorithmException;
+import java.security.Principal;
+import java.security.PublicKey;
+import java.security.SecureRandom;
+import java.security.cert.CertificateException;
+import java.security.cert.CertificateExpiredException;
+import java.security.cert.CertificateNotYetValidException;
+import java.security.cert.X509Certificate;
+import java.util.Calendar;
+import java.util.GregorianCalendar;
+import java.util.Set;
+import java.util.TimeZone;
+import javax.naming.Context;
+import javax.naming.InitialContext;
+import javax.security.auth.Subject;
+import javax.security.auth.x500.X500Principal;
 import org.apache.log4j.Logger;
+import org.bouncycastle.asn1.x500.X500Name;
+import org.bouncycastle.asn1.x509.BasicConstraints;
+import org.bouncycastle.asn1.x509.Extension;
+import org.bouncycastle.asn1.x509.SubjectPublicKeyInfo;
+import org.bouncycastle.cert.X509CertificateHolder;
+import org.bouncycastle.cert.X509v3CertificateBuilder;
+import org.bouncycastle.cert.jcajce.JcaX509CertificateConverter;
+import org.bouncycastle.cert.jcajce.JcaX509ExtensionUtils;
+import org.bouncycastle.openssl.jcajce.JcaPEMWriter;
+import org.bouncycastle.operator.OperatorCreationException;
+import org.bouncycastle.operator.jcajce.JcaContentSignerBuilder;
 
 /**
- * Skeleton action to generate and return a short-lived certificate.
- * 
- * @author pdowler
+ * Class to handle certificate generation requests
+ * @author adriand
  */
 public class GetCertAction extends RestAction {
-    private static final Logger log = Logger.getLogger(GetCertAction.class);
+    protected static Logger log = Logger.getLogger(GetCertAction.class);
+
+    static final String CERTIFICATE_CONTENT_TYPE = "application/x-pem-file";
+    static final String CERTIFICATE_FILENAME = "cadcproxy.pem"; // content disposition
+    static final int CERT_KEY_SIZE = 2048;
+
+    // CADC specific fields of the DN
+    public static final String CADC_DN = "ou=cadc,o=hia,c=ca";
+
+    private CredConfig config;
 
-    public GetCertAction() { 
+    public GetCertAction() {
+        super();
     }
 
     @Override
-    protected InlineContentHandler getInlineContentHandler() {
+    protected final InlineContentHandler getInlineContentHandler() {
         return null;
     }
 
+    @Override
+    public void initAction() {
+        config = CredInitAction.getConfig(super.appName);
+    }
+
     @Override
     public void doAction() throws Exception {
-        throw new UnsupportedOperationException();
+        // create a cert for a single user
+        Subject caller = AuthenticationUtil.getCurrentSubject();
+        String path = syncInput.getPath();
+        if (AuthMethod.CERT.equals(AuthenticationUtil.getAuthMethod(caller)) && !StringUtil.hasText(path)) {
+            throw new AccessControlException("Cert Authentication not allowed for cert renewal.");
```

Comment:
this check could be bypassed by the caller specifying their own identity in the path; it needs to compare caller and the userDN *after* processing the optional path

Reply:
Changed logic to prevent self-renewals with cert authentication.